File: datrisserver/src/main/scala/ai/datris/model/DatasetConfig.scala
package ai.datris.model

/*
Datris
Copyright (C) 2026 Datris (https://datris.ai)
*/

import com.fasterxml.jackson.annotation.{JsonCreator, JsonProperty}

case class DatasetConfig(
                            name: String,
                            source: Source,
                            preprocessor: RestEndpoint,
                            dataQuality: DataQuality,
                            transformation: Transformation,
                            destination: Destination
                        )

case class Source(
                     schemaProperties: SchemaProperties,
                     fileAttributes: FileAttributes,
                     streamAttributes: StreamAttributes,
                     databaseAttributes: DatabaseAttributes
                 )

case class Destination(
                          schemaProperties: SchemaProperties,
                          database: Database,
                          objectStore: ObjectStore,
                          restEndpoint: RestEndpoint,
                          kafka: Kafka,
                          activeMQ: ActiveMQ,
                          qdrant: QdrantConfig,
                          weaviate: WeaviateConfig,
                          pgvector: PGVectorConfig,
                          milvus: MilvusConfig,
                          chroma: ChromaConfig
                      )

case class QdrantConfig(
                           collectionName: String,
                           chunking: ChunkingConfig,
                           metadata: java.util.Map[String, String],
                           embeddingSecretName: String,
                           qdrantSecretName: String
                       )

case class WeaviateConfig(
                             className: String,
                             chunking: ChunkingConfig,
                             metadata: java.util.Map[String, String],
                             embeddingSecretName: String,
                             weaviateSecretName: String
                         )

case class PGVectorConfig(
                             tableName: String,
                             schemaName: String,
                             chunking: ChunkingConfig,
                             metadata: java.util.Map[String, String],
                             embeddingSecretName: String,
                             postgresSecretName: String
                         )

case class MilvusConfig(
                           collectionName: String,
                           chunking: ChunkingConfig,
                           metadata: java.util.Map[String, String],
                           embeddingSecretName: String,
                           milvusSecretName: String
                       )

case class ChromaConfig(
                           collectionName: String,
                           chunking: ChunkingConfig,
                           metadata: java.util.Map[String, String],
                           embeddingSecretName: String,
                           chromaSecretName: String
                       )

case class ChunkingConfig(
                             strategy: String = "recursive",
                             chunkSize: Int = 500,
                             chunkOverlap: Int = 50
                         )

case class SchemaProperties(
                               dbName: String,
                               fields: java.util.List[SchemaField]
                           )

case class DataQuality(
                          validateFileHeader: Boolean,
                          validationSchema: String,
                          aiRule: AIRule,
                          rowRules: java.util.ArrayList[RowRule],
                          columnRules: java.util.List[ColumnRule]
                      )

case class AIRule(
                     instruction: String,
                     onFailureIsError: Boolean,
                     sample: Boolean = false,
                     sampleSize: Int = 200
                 )

case class RowRule(
                      function: String,
                      parameters: java.util.List[String],
                      onFailureIsError: Boolean
                  )

case class ColumnRule(
                         columnName: String,
                         function: String,
                         parameter: String,
                         batchSize: Int = 100,
                         onFailureIsError: Boolean,
                         description: String
                     )

case class Transformation(
                             trimColumnWhitespace: Boolean,
                             deduplicate: Boolean,
                             rowFunctions: java.util.List[RowFunction],
                             aiTransformation: AITransformation
                         )

case class AITransformation(
                               instruction: String,
                               sample: Boolean = false,
                               sampleSize: Int = 200
                           )

case class RowFunction(
                          function: String,
                          parameters: java.util.List[String]
                      )

case class FileAttributes(
                             csvAttributes: CsvAttributes,
                             jsonAttributes: JsonAttributes,
                             xmlAttributes: XmlAttributes,
                             xlsAttributes: XlsAttributes,
                             unstructuredAttributes: UnstructuredAttributes,
                             readOptions: java.util.Map[String, String]
                         )

case class CsvAttributes(
                            delimiter: String,
                            header: Boolean,
                            encoding: String,     // UTF-8, ISO-8859-1, etc]
                        )

case class JsonAttributes(
                             everyRowContainsObject: Boolean,   // If true, each row of the file contains a JSON object
                             encoding: String,     // UTF-8, ISO-8859-1, etc]
                         )

case class XmlAttributes(
                            everyRowContainsObject: Boolean,    // If true, each row of the file contains an XML object
                            encoding: String,     // UTF-8, ISO-8859-1, etc
                        )

case class XlsAttributes(
                            worksheet: Int,
                            tempCsvFileDelimiter: String
                        )

case class UnstructuredAttributes(
                                     fileExtension: String,
                                     preserveFilename: Boolean
                                 )

case class StreamAttributes(
                               `type`: String
                           )

case class DatabaseAttributes(
                                 `type`: String,
                                 postgresSecretsName: String,
                                 mssqlSecretsName: String,
                                 mysqlSecretsName: String,
                                 cronExpression: String,
                                 database: String,
                                 schema: String,
                                 table: String,
                                 includeFields: java.util.List[String],
                                 timestampFieldName: String,
                                 sqlOverride: String,
                                 outputDelimiter: String
                             )
case class RestEndpoint(
                           endpoint: String,
                           async: Boolean = false,
                           bearerToken: String = null,
                           timeoutSeconds: Int = 300
                       )
case class ObjectStore(
                          prefixKey: String,
                          partitionBy: java.util.List[String],
                          destinationBucketOverride: String,
                          fileFormat: String,
                          writeToTemporaryLocation: Boolean,
                          deleteBeforeWrite: Boolean,
                          writeMode: String
                      )

case class Database(
                       dbName: String, // Database name
                       schema: String,
                       table: String, // Table name
                       keyFields: java.util.List[String],
                       manageTableManually: Boolean,
                       truncateBeforeWrite: Boolean,
                       useTransaction: Boolean,
                       usePostgres: Boolean,
                       useMongoDB: Boolean,
                       options: java.util.List[String]
                   )

case class Kafka(
                    topic: String,
                    keyField: String,
                    overrideBootstrapServers: String,
                    timeoutMs: Int = 10000
                )

case class ActiveMQ @JsonCreator() (
                    @JsonProperty("queueName") queueName: String
                    )

File: ui/src/model/dataset_config.py
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class CsvAttributes:
    delimiter: Optional[str] = None
    header: Optional[bool] = None
    encoding: Optional[str] = None  # UTF-8, ISO-8859-1, etc


@dataclass
class JsonAttributes:
    every_row_contains_object: Optional[bool] = None  # If true, each row contains a JSON object
    encoding: Optional[str] = None  # UTF-8, ISO-8859-1, etc


@dataclass
class XmlAttributes:
    every_row_contains_object: Optional[bool] = None  # If true, each row contains an XML object
    encoding: Optional[str] = None  # UTF-8, ISO-8859-1, etc


@dataclass
class XlsAttributes:
    worksheet: Optional[int] = None
    temp_csv_file_delimiter: Optional[str] = None


@dataclass
class UnstructuredAttributes:
    file_extension: Optional[str] = None
    preserve_filename: Optional[bool] = None


@dataclass
class DatabaseAttributes:
    type: Optional[str] = None
    postgres_secrets_name: Optional[str] = None
    mssql_secrets_name: Optional[str] = None
    mysql_secrets_name: Optional[str] = None
    cron_expression: Optional[str] = None
    database: Optional[str] = None
    schema: Optional[str] = None
    table: Optional[str] = None
    include_fields: list[str] = field(default_factory=list)
    timestamp_field_name: Optional[str] = None
    sql_override: Optional[str] = None
    output_delimiter: Optional[str] = None


@dataclass
class FileAttributes:
    csv_attributes: Optional[CsvAttributes] = None
    json_attributes: Optional[JsonAttributes] = None
    xml_attributes: Optional[XmlAttributes] = None
    xls_attributes: Optional[XlsAttributes] = None
    unstructured_attributes: Optional[UnstructuredAttributes] = None
    spark_read_options: dict[str, str] = field(default_factory=dict)


@dataclass
class SchemaField:
    name: Optional[str] = None
    type: Optional[str] = None


@dataclass
class SchemaProperties:
    db_name: Optional[str] = None
    fields: list[SchemaField] = field(default_factory=list)


@dataclass
class Source:
    schema_properties: Optional[SchemaProperties] = None
    file_attributes: Optional[FileAttributes] = None
    database_attributes: Optional[DatabaseAttributes] = None


@dataclass
class Database:
    db_name: Optional[str] = None
    schema: Optional[str] = None
    table: Optional[str] = None
    key_fields: list[str] = field(default_factory=list)
    manage_table_manually: Optional[bool] = None
    truncate_before_write: Optional[bool] = None
    use_postgres: Optional[bool] = None
    options: list[str] = field(default_factory=list)


@dataclass
class ObjectStore:
    prefix_key: Optional[str] = None
    partition_by: list[str] = field(default_factory=list)
    destination_bucket_override: Optional[str] = None
    file_format: Optional[str] = None
    write_to_temporary_location: Optional[bool] = None
    delete_before_write: Optional[bool] = None
    manage_glue_table_manually: Optional[bool] = None
    use_iceberg: Optional[bool] = None
    key_fields: list[str] = field(default_factory=list)
    use_athena: Optional[bool] = None
    use_spark_cluster: Optional[bool] = None
    spark_write_mode: Optional[str] = None


@dataclass
class Destination:
    schema_properties: Optional[SchemaProperties] = None
    database: Optional[Database] = None
    object_store: Optional[ObjectStore] = None


@dataclass
class DataQuality:
    validate_file_header: Optional[bool] = None
    validation_schema: Optional[str] = None


@dataclass
class Transformation:
    ai_transformation_instruction: Optional[str] = None


@dataclass
class DatasetConfig:
    name: str
    source: Source
    data_quality: Optional[DataQuality] = None
    transformation: Optional[Transformation] = None
    destination: Destination